backend/src/main/java/vn/edu/hust/nmcnpm_20242_n3/service/SubscriptionService.java

```diff
@@ -0,0 +1,186 @@
+import vn.edu.hust.nmcnpm_20242_n3.constant.BookCopyStatusEnum;
+import vn.edu.hust.nmcnpm_20242_n3.entity.BookCopy;
+import vn.edu.hust.nmcnpm_20242_n3.entity.Subscription;
+import vn.edu.hust.nmcnpm_20242_n3.repository.BookCopyRepository;
+import vn.edu.hust.nmcnpm_20242_n3.repository.SubscriptionRepository;
+import vn.edu.hust.nmcnpm_20242_n3.repository.UserRepository;
+
+import java.util.List;
+import java.util.Optional;
+
+@Service
+public class SubscriptionService {
+
+    private final BookCopyRepository bookCopyRepository;
+    private final SubscriptionRepository subscriptionRepository;
+    private final UserRepository userRepository;
+    private final JavaMailSender emailSender;
+
+    @Autowired
+    private BookRequestService bookRequestService;
+
+    public SubscriptionService(BookCopyRepository bookCopyRepository, SubscriptionRepository subscriptionRepository,
+                               UserRepository userRepository, JavaMailSender mailSender ) {
+        this.bookCopyRepository = bookCopyRepository;
+        this.subscriptionRepository = subscriptionRepository;
+        this.userRepository = userRepository;
+        this.emailSender = mailSender; // Initialize with your email sender bean
+    }
+
+    public void subscribeToBookCopy(int bookCopyId, String userId) {
+        // Check if the book copy exists
+        BookCopy bookCopy = bookCopyRepository.findById(bookCopyId)
+                .orElseThrow(() -> new IllegalArgumentException("Book copy not found"));
+
+        // Create a new subscription
+        Subscription subscription = new Subscription();
+        subscription.setBookCopy(bookCopy);
+        subscription.setUser(userRepository.findById(userId)
+                .orElseThrow(() -> new IllegalArgumentException("User not found")));
+        subscription.setActive(true);
+
+        // Save the subscription
+        subscriptionRepository.save(subscription);
+    }
+
+    public void unsubscribeFromBookCopy(int bookCopyId, String userId) {
+        // Find the subscription
+        Subscription subscription = subscriptionRepository.findByBookCopyIdAndUserId(bookCopyId, userId)
+                .orElseThrow(() -> new IllegalArgumentException("Subscription not found"));
+
+        // Mark the subscription as inactive
+        subscription.setActive(false);
+        subscriptionRepository.save(subscription);
+    }
+
+    public void notifyUsers(int bookCopyId) {
+        // Get all active subscriptions for the book copy
+        List<Subscription> subscriptions = subscriptionRepository.findAllByBookCopyIdAndActive(bookCopyId, true);
+
+        // Notify each user
+        for (Subscription subscription : subscriptions) {
+            String email = subscription.getUser().getEmail();
+            this.sendEmail(email, "Book Copy Available", "The book copy you subscribed to is now available.");
+            System.out.println("Notification sent to: " + email);
+        }
+    }
```

Comment:
The content of the email is kinda bland. Should we perhaps adding more information, like the name of the book and its copy id, so the email would be more dynamic?